Rewrite the dual-image dataset and engine

cshuoyu opened the pull request:
This PR introduces the new dual-image dataset and engine.

First, I added support for an extended HDF5 file structure.
The previous HDF5 file structure is like:
angles | Shape: (5035498, 2) | Dtype: float32
energies | Shape: (5035498, 1) | Dtype: float32
event_hits_index | Shape: (5035498,) | Dtype: int64
event_ids | Shape: (5035498,) | Dtype: int32
hit_charge | Shape: (18000976589,) | Dtype: float32
hit_pmt | Shape: (18000976589,) | Dtype: int32
hit_time | Shape: (18000976589,) | Dtype: float32
here we just have 20 inch PMTs' data.

To integrate mPMTs' data, now I'm using HDF5 file structure like:
Group: main
    angles | Shape: (5228031, 2) | Dtype: float32
    energies | Shape: (5228031, 1) | Dtype: float32
    event_hits_index | Shape: (5228031,) | Dtype: int64
    event_ids | Shape: (5228031,) | Dtype: int32
    hit_charge | Shape: (18688388735,) | Dtype: float32
    hit_pmt | Shape: (18688388735,) | Dtype: int32
    hit_time | Shape: (18688388735,) | Dtype: float32
    labels | Shape: (5228031,) | Dtype: int32
    positions | Shape: (5228031, 1, 3) | Dtype: float32
    root_event_id | Shape: (5228031,) | Dtype: int32
    root_files | Shape: (5228031,) | Dtype: object
Group: mpmt
    angles | Shape: (5228031, 2) | Dtype: float32
    energies | Shape: (5228031, 1) | Dtype: float32
    event_hits_index | Shape: (5228031,) | Dtype: int64
    event_ids | Shape: (5228031,) | Dtype: int32
    hit_charge | Shape: (568277176,) | Dtype: float32
    hit_pmt | Shape: (568277176,) | Dtype: int32
    hit_time | Shape: (568277176,) | Dtype: float32
    labels | Shape: (5228031,) | Dtype: int32
    positions | Shape: (5228031, 1, 3) | Dtype: float32
    root_event_id | Shape: (5228031,) | Dtype: int32
    root_files | Shape: (5228031,) | Dtype: object

In this PR, I added this structure into h5data and the previous structure is still compatible to use.

Moreover, I rewrite the dual-image engine.
I've tested that I can run these engine and dataset in a dual-image model.

cshuoyu committed:
Rewrite the dual-image dataset and engine

cshuoyu committed:
Update the Vision Transformer models

cshuoyu commented:
I just updated the vision transformer models in watchmal/model.

1. I deleted T2TViT models and ViT model.
2. I changed previous SwinT model which was implemented by timm into an example for how to use timm
3. I add swint single image and dual image model implemented from scratch.

I've basically tested they can run well.

nickwp commented on watchmal/engine/regression.py:
```diff
@@ -93,3 +93,36 @@ def restore_state(self, weight_file):
+        data_main, data_second = data["data"]
+        self.data_main = data_main.to(self.device)
+        self.data_second = data_second.to(self.device)
+        self.data = (self.data_main, self.data_second)
```
I think the existing process_data already does this [here](https://github.com/WatChMaL/WatChMaL/blob/924dd5403f3ea2c6fea744f869238304f15f81fb/watchmal/engine/reconstruction.py#L152-L157).

nickwp reviewed (changes requested):
Hi @cshuoyu 
Do you have any update regarding the comments I made and the code suggestions? Let me know if you need anything more if you are able to continue with this work. It would be very nice if we could update to merge and include these changes

cshuoyu committed:
Add pmt_type for H5 datasets

cshuoyu committed:
Delete RegressionDIEngine for DI models

Erwanlbv commented:
Hello, 
i'd like to implement this work as soon as possible in the merge. 
Do you have any status ?

nickwp commented on config/engine/regression_DI.yaml:
Since the RegressionDIEngine was removed, is this needed?
It seems there is nothing special about this config that means it should be specific to DI, so it seems a bit misleading to have this engine config for DI.
It seems more like a config just for general energy regression.
Maybe it can be removed?

cshuoyu replied:
Yes, I think so.

nickwp commented:
> Hello, i'd like to implement this work as soon as possible in the merge. Do you have any status ?

This needs to be tested properly before it can be merged, but unfortunately I haven't yet had time to set up an environment for testing it. I'll try to do so as soon as possible.

But since this basically only affects the CNN dataset code and model code, is it relevant for the merge?

Erwanlbv commented:
I'd like to compare ViT and "Vit-style" GNN as soon as possible on the same MC, I think I'd be better to do it with same code as Shuoyu (even thought I could write myself the ViT part)

I got access to compute canada so I should be able to train my models there

nickwp commented:
> I'd like to compare ViT and "Vit-style" GNN as soon as possible on the same MC, I think I'd be better to do it with same code as Shuoyu (even thought I could write myself the ViT part)
> 
> I got access to compute canada so I should be able to train my models there

Okay, I'll try to test this so it can be merged asap. But it might have to be after the HK and WCTE CMs because of so many other priorities right now

cshuoyu commented:
> I'd like to compare ViT and "Vit-style" GNN as soon as possible on the same MC, I think I'd be better to do it with same code as Shuoyu (even thought I could write myself the ViT part)
> 
> I got access to compute canada so I should be able to train my models there

After we merge this PR, we can have the SwinT model. But one more thing is I still haven't merged the optimizer that I'm using for SwinT ( I replace AdamW with Muon https://github.com/KellerJordan/Muon, I found this one has a faster convergence). I tried to do that, but when I tested that, I found it cannot work for single GPU (basically I used multi GPUs for training), and I didn't have time to solve that problem that time. 

I think if we are going to compare them on the same MC, we'd better have this optimizer, because my hyperparameters are based on this. I'm trying to merge this optimizer soon.

cshuoyu committed:
Update: multi-task for regression

cshuoyu committed:
Update: muon optimizer

cshuoyu committed:
Feature: add mPMT geometry position to swintDI

cshuoyu committed:
Update: clear up the swint config

cshuoyu commented:
Hi, sorry for the delay. I just submitted four commits. Let me explain them:

1.  [Update: multi-task for regression](https://github.com/WatChMaL/WatChMaL/pull/97/commits/5734cfc5f42df513f9f0d6dfc4ed5f9ae7ee43c1) 

I found I didn't merge the multi-task method, so I updated them like in this commit. The method is like the one I mentioned in several meeting (trainable weight).

2.  [Update: muon optimizer](https://github.com/WatChMaL/WatChMaL/pull/97/commits/b8df34c5e628264fdb37c8c1db2ec3f8bee7bbbf)

I'm sorry I delay this feature for a very long time. Here is a new optimizer call Muon, which shows much better performance than adam.  [It's widely used in LLM training now](https://arxiv.org/abs/2502.16982v1), so it indeed can help a transformer-based model It's difficult to say whether it can help the final model performance, but it's very obvious that the convergence is faster, so I used muon as my default optimizer for all the results I showed.

https://github.com/KellerJordan/Muon. I have some problems when I implemented muon in watchmal. I initially tried installing their library via pip, but I found that in that setup I could not run my tests on a single GPU or on CPU.
My current solution is to vendor muon.py directly into watchmal/utils and adapt the integration so it works correctly in single-device and CPU settings. With that change, it now runs normally.
I am not fully sure whether this is the right or acceptable way to handle it.

3. [Feature: add mPMT geometry position to swintDI](https://github.com/WatChMaL/WatChMaL/pull/97/commits/677e3cd0dbe837f590a56ea3a6cb3ec471d2c14b)

Recently, I added a new feature to swintDI model, which can integrate the real geometry information of the mPMT into the model.

4. [Update: clear up the swint config](https://github.com/WatChMaL/WatChMaL/pull/97/commits/3bb54d86c7205c2fece9b7ccbf834796590c640b)

Some configuration for swint

after these commits, in principle it can run in the same way as my local one.

cshuoyu committed:
clean: multi-task

nickwp commented on watchmal/dataset/h5_dataset.py:
```diff
@@ -169,14 +169,28 @@ def __getitem__(self, item):
+
+        if self.pmt_type is not None:
+            if event_hit_pmt_types is None:
+                print(f"WARNING: 'hit_pmt_type' not found in {self.h5_path}; pmt_type filter ignored")
```
Could you use the built in logging functionality with `log.warning` instead of a direct print? I think it will need to be imported because this file doesn't yet have any other warnings or logging

nickwp commented on watchmal/dataset/h5_dataset.py:
```diff
@@ -228,12 +242,27 @@ def __getitem__(self, item):
+
+        if self.pmt_type is not None:
+            if true_pmt_types is None:
+                print(f"WARNING: 'hit_pmt_type' not found in {self.h5_path}; pmt_type filter ignored")
```
As the other comment above, about using logging instead of print

nickwp commented on watchmal/dataset/dualimage/dualimage_dataset.py:
```diff
@@ -0,0 +1,94 @@
+        hit_pmts = self.hit_pmt[start:stop]
+        hit_times = self.hit_time[start:stop]
+        hit_charges = self.hit_charge[start:stop]
+        hit_pmt_types = self.hit_pmt_type[start:stop] if self.hit_pmt_type is not None else None
+
+        if self.pmt_type_mpmt is not None:
+            if hit_pmt_types is None:
+                print(f"WARNING: 'hit_pmt_type' not found in {self.h5_path}; mPMT hits set to empty")
+                hit_pmts = hit_pmts[:0]
+                hit_times = hit_times[:0]
+                hit_charges = hit_charges[:0]
+            else:
```
it's not too important, but I think it's cleaner and easier to read what's happening if this assignment is moved below the else where it's used, and so the if else condition is not repeated in this line and also below


```suggestion
        if self.pmt_type_mpmt is not None:
            if self.hit_pmt_type is None:
                print(f"WARNING: 'hit_pmt_type' not found in {self.h5_path}; mPMT hits set to empty")
                hit_pmts = hit_pmts[:0]
                hit_times = hit_times[:0]
                hit_charges = hit_charges[:0]
            else:
                hit_pmt_types = self.hit_pmt_type[start:stop]
```

and similarly for the two other places this pattern is used in this PR

nickwp commented on watchmal/dataset/dualimage/dualimage_dataset.py:
```diff
@@ -0,0 +1,94 @@
+
+        if self.pmt_type_mpmt is not None:
+            if hit_pmt_types is None:
+                print(f"WARNING: 'hit_pmt_type' not found in {self.h5_path}; mPMT hits set to empty")
```
Could you use the built in logging functionality with log.warning instead of a direct print? I think it will need to be imported because this file doesn't yet have any other warnings or logging
